src/VisualStudio/Core/Def/DocumentOutline/VisualStudioCodeWindowInfoService.cs

```diff
@@ -0,0 +1,134 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.TextManager.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    /// <summary>
+    /// Manages access to Visual Studio code window information in a thread-safe way.
+    /// </summary>
+    internal class VisualStudioCodeWindowInfoService
+    {
+        private readonly IVsCodeWindow _codeWindow;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly IThreadingContext _threadingContext;
+
+        public VisualStudioCodeWindowInfoService(
+            IVsCodeWindow codeWindow,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IThreadingContext threadingContext)
+        {
+            threadingContext.ThrowIfNotOnUIThread();
+            _codeWindow = codeWindow;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _threadingContext = threadingContext;
+        }
+
+        /// <summary>
+        /// Get <see cref="VisualStudioCodeWindowInfo"/>, switching to the UI thread if necessary.
+        /// </summary>
+        public async ValueTask<VisualStudioCodeWindowInfo?> GetVisualStudioCodeWindowInfoAsync(CancellationToken token)
+        {
+            if (!_threadingContext.JoinableTaskContext.IsOnMainThread)
+            {
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(token);
+            }
+
+            var wpfTextView = GetLastActiveIWpfTextView();
+            if (wpfTextView is null)
+            {
+                return null;
+            }
+            var textBuffer = wpfTextView.TextBuffer;
+
+            var caretPoint = wpfTextView.GetCaretPoint(textBuffer);
+
+            if (_editorAdaptersFactoryService.GetBufferAdapter(textBuffer) is IPersistFileFormat persistFileFormat &&
+                ErrorHandler.Succeeded(persistFileFormat.GetCurFile(out var filePath, out var _)))
+            {
+                return new VisualStudioCodeWindowInfo(textBuffer, filePath, caretPoint);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Returns a version of this service whose apis can only be called on the UI thread.
+        /// </summary>
+        public VisualStudioCodeWindowInfoService_OnlyCallOnUIThread GetServiceAndThrowIfNotOnUIThread()
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+            return new VisualStudioCodeWindowInfoService_OnlyCallOnUIThread(this);
+        }
+
+        private SnapshotPoint? GetCurrentCaretSnapshotPoint()
+        {
```

Comment:
can you add threadin asserts to all these methods?

Reply:
yep